hazelcast/src/main/java/com/hazelcast/query/impl/Indexes.java

```diff
@@ -204,6 +205,53 @@ public void createIndexesFromRecordedDefinitions(StoreAdapter partitionStoreAdap
         return indexes;
     }
 
+    /**
+     * Returns known and populated indexes for the given partition. If no
+     * indexes are known or populated, empty array is returned.
+     * <p>
+     * Must be called on partition threads only.
+     */
+    public InternalIndex[] getPopulatedIndexes(int partitionId) {
+        ThreadUtil.assertRunningOnPartitionThread();
+
+        InternalIndex[] indexes = getIndexes();
+
+        int populatedCount = 0;
+        for (InternalIndex index : indexes) {
+            if (index.hasPartitionIndexed(partitionId)) {
+                ++populatedCount;
+            }
+        }
+
+        if (populatedCount == indexes.length) {
+            return indexes;
+        }
+        if (populatedCount == 0) {
+            return EMPTY_INDEXES;
+        }
+
+        // The code below handles a rare case which may happen during indexes
+        // creation/population. E.g., see MapProxySupport.initializeIndexes:
+        // indexes are added one by one, for every index AddIndexOperation is
+        // issued on all partitions. Individual add index operations may
+        // interleave in any order with any operations including add index
+        // operations issued by other members. So for a global index only some
+        // of its associated partitions might be indexed, but that should not
+        // happen for partitioned HD indexes: all known indexes must be either
```

Comment:
> but that should not happen for partitioned HD indexes

How is this so? How do we achieve that it's all-or-nothing? Because we populate on the partition thread?

Reply:
Yes, because (1) we populate indexes from partition threads and (2) HD indexes are managed on a per-partition basis, so all HD indexes known to a certain partition must be either populated or not populated as observed from the associated partition thread at any given time; while for global indexes some indexes might be created, but not populated yet, as a result of creating and populating the indexes for some other partition on some other partition thread.